src/internal/packager/images/pull.go

```diff
@@ -27,329 +32,329 @@ import (
 	"github.com/google/go-containerregistry/pkg/v1/empty"
 	clayout "github.com/google/go-containerregistry/pkg/v1/layout"
 	"github.com/google/go-containerregistry/pkg/v1/partial"
-	ctypes "github.com/google/go-containerregistry/pkg/v1/types"
+	"github.com/google/go-containerregistry/pkg/v1/remote"
+	"github.com/google/go-containerregistry/pkg/v1/types"
 	"github.com/moby/moby/client"
+	ocispec "github.com/opencontainers/image-spec/specs-go/v1"
+	"golang.org/x/sync/errgroup"
 )
 
 // ImgInfo wraps references/information about an image
 type ImgInfo struct {
-	RefInfo        transform.Image
-	Img            v1.Image
-	HasImageLayers bool
+	RefInfo transform.Image
+	Img     v1.Image
 }
 
-// PullAll pulls all of the images in the provided tag map.
-func (i *ImageConfig) PullAll() ([]ImgInfo, error) {
-	var (
-		longer            string
-		imageCount        = len(i.ImageList)
-		refInfoToImage    = map[transform.Image]v1.Image{}
-		referenceToDigest = make(map[string]string)
-		imgInfoList       []ImgInfo
-	)
-
-	type digestInfo struct {
-		refInfo transform.Image
-		digest  string
-	}
-
+// Pull pulls all of the images from the given config.
+func Pull(ctx context.Context, cancel context.CancelFunc, cfg PullConfig) (map[transform.Image]v1.Image, error) {
+	var longer string
+	imageCount := len(cfg.References)
 	// Give some additional user feedback on larger image sets
 	if imageCount > 15 {
 		longer = "This step may take a couple of minutes to complete."
 	} else if imageCount > 5 {
 		longer = "This step may take several seconds to complete."
 	}
 
-	spinner := message.NewProgressSpinner("Loading metadata for %d images. %s", imageCount, longer)
+	if err := helpers.CreateDirectory(cfg.DestinationDirectory, helpers.ReadExecuteAllWriteUser); err != nil {
+		return nil, fmt.Errorf("failed to create image path %s: %w", cfg.DestinationDirectory, err)
+	}
+
+	cranePath, err := clayout.FromPath(cfg.DestinationDirectory)
+	if err != nil {
+		cranePath, err = clayout.Write(cfg.DestinationDirectory, empty.Index)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	spinner := message.NewProgressSpinner("Fetching info for %d images. %s", imageCount, longer)
 	defer spinner.Stop()
 
 	logs.Warn.SetOutput(&message.DebugWriter{})
 	logs.Progress.SetOutput(&message.DebugWriter{})
 
-	metadataImageConcurrency := helpers.NewConcurrencyTools[ImgInfo, error](len(i.ImageList))
+	eg, _ := errgroup.WithContext(ctx)
+	eg.SetLimit(10)
 
-	defer metadataImageConcurrency.Cancel()
+	var fetchedLock, shaLock sync.Mutex
+	shas := make(map[string]bool)
```

Comment:
NIT: Keep declaration syntax the same.

```suggestion
	shas := map[string]bool{}
```